src/panic.rs

```diff
@@ -0,0 +1,69 @@
+use crate::error::rustls_result;
+
+use std::ptr::{null, null_mut};
+
+// We wrap all function calls in an ffi_panic_boundary! macro, which catches
+// panics and early-returns from the function. For functions that return
+// rustls_result, we return a dedicated error code: `Panic`. For functions
+// that don't return rustls_result, we return a default value: false, 0, or
+// null. This trait provides that logic.
+// Note: It's tempting to do a blanket `impl<T> PanicOrDefault for T: Default`.
+// However, that would conflict with the impls for `*mut T` and `*const T`.
+// Though those types currently don't implement Default, Rust disallows the
+// conflict because upstream might later implement it for them.
+pub(crate) trait PanicOrDefault {
+    fn value() -> Self;
+}
+
+impl PanicOrDefault for rustls_result {
+    fn value() -> Self {
+        rustls_result::Panic
+    }
+}
+
+impl PanicOrDefault for u16 {
+    fn value() -> Self {
+        Default::default()
+    }
```

Comment:
nit: To reduce repetition, you could hoist this implementation of `value()` up into a default implementation on `trait PanicOrDefault`, and then the `impl`s for `u16`, `usize`, `bool` and `()` are reduced to opting into `PanicOrDefault`.